File: generators/chords.py
from __future__ import annotations

from music21 import chord, interval, meter, note, stream

from models import NOTE_VALUES
from utils import parse_pitch, validate_range


CHORD_INTERVALS = {
    "Major triad": ["P1", "M3", "P5"],
    "Minor triad": ["P1", "m3", "P5"],
    "Diminished triad": ["P1", "m3", "d5"],
    "Augmented triad": ["P1", "M3", "A5"],
    "Dominant 7": ["P1", "M3", "P5", "m7"],
    "Major 7": ["P1", "M3", "P5", "M7"],
    "Minor 7": ["P1", "m3", "P5", "m7"],
}


def _find_root_octave(root: str, low: str, high: str, octave_span: int) -> int | None:
    low_p = parse_pitch(low)
    high_p = parse_pitch(high)
    for octave in range(2, 8):
        start = parse_pitch(f"{root}{octave}")
        end = parse_pitch(f"{root}{octave + octave_span}")
        if start.midi >= low_p.midi and end.midi <= high_p.midi:
            return octave
    return None


def _build_tones(root_pitch, chord_type: str):
    tones = []
    for iv in CHORD_INTERVALS[chord_type]:
        tones.append(interval.Interval(iv).transposePitch(root_pitch))
    return tones


def build_chord_exercise(settings: dict) -> tuple[stream.Score | None, str | None, str | None]:
    err = validate_range(settings["lowest_note"], settings["highest_note"])
    if err:
        return None, None, err

    base_oct = _find_root_octave(settings["root"], settings["lowest_note"], settings["highest_note"], settings["octave_span"])
    if base_oct is None:
        return None, None, "This chord pattern cannot be generated within the chosen range."

    score = stream.Score()
    part = stream.Part()
    part.append(meter.TimeSignature(settings["time_signature"]))

    base_root = parse_pitch(f"{settings['root']}{base_oct}")
    tones = _build_tones(base_root, settings["chord_type"])

    low_midi = parse_pitch(settings["lowest_note"]).midi
    high_midi = parse_pitch(settings["highest_note"]).midi

    pattern = settings["pattern"]
    ql = NOTE_VALUES[settings["note_value"]]

    if pattern == "Block chord":
        for i in range(settings["octave_span"]):
            octave_tones = [p.transpose(12 * i) for p in tones]
            if not all(low_midi <= p.midi <= high_midi for p in octave_tones):
                return None, None, "This chord pattern cannot be generated within the chosen range."
            c = chord.Chord(octave_tones)
            c.quarterLength = ql
            if settings["show_chord_tones"]:
                c.lyric = " ".join([p.name for p in octave_tones])
            part.append(c)
    else:
        seq = []
        for i in range(settings["octave_span"]):
            octave_tones = [p.transpose(12 * i) for p in tones]
            seq.extend(octave_tones)
        seq.append(base_root.transpose(12 * settings["octave_span"]))

        if pattern == "Descending arpeggio":
            seq = list(reversed(seq))
        elif pattern == "Up and Down":
            seq = seq + list(reversed(seq))[1:]

        for p in seq:
            if not (low_midi <= p.midi <= high_midi):
                return None, None, "This chord pattern cannot be generated within the chosen range."
            n = note.Note(p)
            n.quarterLength = ql
            if settings["show_note_names"] or settings["show_chord_tones"]:
                lbls: list[str] = []
                if settings["show_note_names"]:
                    lbls.append(n.nameWithOctave)
                if settings["show_chord_tones"]:
                    lbls.append(n.name)
                n.lyric = " / ".join(lbls)
            part.append(n)

    score.append(part)

    chord_name = settings["chord_type"].replace(" triad", "")
    if "7" in settings["chord_type"]:
        chord_name = settings["chord_type"].replace("Dominant 7", "7")
    title = f"{settings['root']} {chord_name} Arpeggio - {settings['octave_span']} Octave - {settings['pattern']}"

    return score, title, None

File: src/lib/generators/chords.ts
import { Exercise, GenerateResult, MusicEvent } from "@/lib/music/models";
import { buildChordPitchClasses, parsePitch } from "@/lib/music/noteUtils";
import { chunkIntoMeasures, uid } from "@/lib/generators/shared";
import { ChordSettings, validateChordSettings } from "@/lib/validation/chordsValidation";

const CHORD_INTERVALS: Record<string, number[]> = {
  "Major triad": [0, 4, 7],
  "Minor triad": [0, 3, 7],
  "Diminished triad": [0, 3, 6],
  "Augmented triad": [0, 4, 8],
  "Dominant 7": [0, 4, 7, 10],
  "Major 7": [0, 4, 7, 11],
  "Minor 7": [0, 3, 7, 10]
};

const findRootOctave = (root: string, low: string, high: string, octaveSpan: number): number | undefined => {
  const lowMidi = parsePitch(low).midi;
  const highMidi = parsePitch(high).midi;
  for (let octave = 2; octave <= 7; octave += 1) {
    const start = parsePitch(`${root}${octave}`).midi;
    const end = parsePitch(`${root}${octave + octaveSpan}`).midi;
    if (start >= lowMidi && end <= highMidi) return octave;
  }
  return undefined;
};

export const generateChordExercise = (settings: ChordSettings): GenerateResult => {
  const err = validateChordSettings(settings);
  if (err) return { error: err };

  const intervals = CHORD_INTERVALS[settings.chordType];
  if (!intervals) return { error: "Unsupported chord type." };

  const rootOct = findRootOctave(settings.root, settings.lowestNote, settings.highestNote, settings.octaveSpan);
  if (rootOct === undefined) return { error: "This arpeggio pattern does not fit the selected range." };

  const lowMidi = parsePitch(settings.lowestNote).midi;
  const highMidi = parsePitch(settings.highestNote).midi;
  const template = buildChordPitchClasses(settings.root, settings.chordType, intervals);
  const events: MusicEvent[] = [];

  if (settings.pattern === "Block chord") {
    for (let oct = 0; oct < settings.octaveSpan; oct += 1) {
      const chordTones = template.map((degree) => `${degree.pitchClass}${rootOct + oct + degree.octaveOffset}`);
      if (chordTones.some((pitch) => {
        const midi = parsePitch(pitch).midi;
        return midi < lowMidi || midi > highMidi;
      })) {
        return { error: "This arpeggio pattern does not fit the selected range." };
      }
      events.push({
        kind: "note",
        pitch: chordTones[0],
        chord: chordTones,
        duration: settings.noteValue,
        lyric: settings.showChordTones ? chordTones.map((tone) => tone.replace(/\d/, "")).join(" ") : undefined
      });
    }
  } else {
    const seq: string[] = [];
    for (let oct = 0; oct < settings.octaveSpan; oct += 1) {
      for (const degree of template) seq.push(`${degree.pitchClass}${rootOct + oct + degree.octaveOffset}`);
    }
    seq.push(`${template[0].pitchClass}${rootOct + settings.octaveSpan}`);

    const run = settings.pattern === "Descending arpeggio"
      ? [...seq].reverse()
      : settings.pattern === "Up and Down"
      ? [...seq, ...[...seq].reverse().slice(1)]
      : seq;

    for (const pitch of run) {
      const midi = parsePitch(pitch).midi;
      if (midi < lowMidi || midi > highMidi) return { error: "This arpeggio pattern does not fit the selected range." };
      const labels: string[] = [];
      if (settings.showNoteNames) labels.push(pitch);
      if (settings.showChordTones) labels.push(pitch.replace(/\d/, ""));
      events.push({ kind: "note", pitch, duration: settings.noteValue, lyric: labels.length ? labels.join(" / ") : undefined });
    }
  }

  const label = settings.pattern === "Block chord" ? "Chord" : "Arpeggio";

  const exercise: Exercise = {
    id: uid(),
    type: "chord",
    title: `${settings.root} ${settings.chordType} ${label} - ${settings.octaveSpan} Octave - ${settings.pattern}`,
    timeSignature: settings.timeSignature,
    clef: "treble",
    metadata: {
      root: settings.root,
      chordType: settings.chordType,
      pattern: settings.pattern,
      octaveSpan: settings.octaveSpan,
      range: `${settings.lowestNote}-${settings.highestNote}`
    },
    measures: chunkIntoMeasures(events, settings.timeSignature)
  };

  return { exercise };
};
